File: src/endpoint/Bc.Endpoint/GitHubPullRequestCreation.cs
using System;
using System.Threading.Tasks;
using Bc.Contracts.Internals.Endpoint.GitHubPullRequestCreation;
using Bc.Contracts.Internals.Endpoint.GitHubPullRequestCreation.Messages;
using NServiceBus;
using NServiceBus.Persistence.Sql;

namespace Bc.Endpoint.GitHubPullRequestCreation
{
    //[SqlSaga(tableSuffix: "GitHubPullRequestCreationPolicy")]
    public class Policy :
        Saga<Policy.PolicyData>,
        IAmStartedByMessages<RequestCreateGitHubPullRequest>,
        IHandleMessages<ResponseCreateBranch>,
        IHandleMessages<ResponseUpdateFile>,
        IHandleMessages<ResponseCreatePullRequest>
    {
        public Task Handle(RequestCreateGitHubPullRequest message, IMessageHandlerContext context)
        {
            this.Data.UserName = message.UserName;
            this.Data.UserWebSite = message.UserWebsite;
            this.Data.FileName = message.FileName;
            this.Data.Content = message.Content;
            this.Data.AddedDate = message.AddedDate;

            return context.Send(new RequestCreateBranch(this.Data.AddedDate));
        }

        public Task Handle(ResponseCreateBranch message, IMessageHandlerContext context)
        {
            this.Data.BranchName = message.BranchName;
            return context.Send(new RequestUpdateFile(this.Data.BranchName, this.Data.FileName, this.Data.Content));
        }

        public Task Handle(ResponseUpdateFile message, IMessageHandlerContext context)
        {
            return context.Send(new RequestCreatePullRequest(this.Data.BranchName));
        }

        public Task Handle(ResponseCreatePullRequest message, IMessageHandlerContext context)
        {
            this.MarkAsComplete();
            return this.ReplyToOriginator(
                context,
                new ResponseCreateGitHubPullRequest(this.Data.CommentId, message.PullRequestUri));
        }
        
        protected override void ConfigureHowToFindSaga(SagaPropertyMapper<PolicyData> mapper)
        {
            mapper.ConfigureMapping<RequestCreateGitHubPullRequest>(message => message.CommentId)
                  .ToSaga(data => data.CommentId);
        }        

        public class PolicyData : ContainSagaData
        {
            public Guid CommentId { get; set; }

            public string UserName { get; set; }

            public string UserWebSite { get; set; }

            public string FileName { get; set; }

            public string Content { get; set; }

            public DateTime AddedDate { get; set; }

            public string BranchName { get; set; }    
        }
    }
    
    public class PolicyHandlers :
        IHandleMessages<RequestCreateBranch>,
        IHandleMessages<RequestUpdateFile>,
        IHandleMessages<RequestCreatePullRequest>
    {
        private readonly IPolicyLogic logic;

        public PolicyHandlers(IPolicyLogic logic)
        {
            this.logic = logic;
        }        
        
        public async Task Handle(RequestCreateBranch message, IMessageHandlerContext context)
        {
            var branchName = await this.logic.CreateBranch(message.AddedDate).ConfigureAwait(false);
            await context.Reply(new ResponseCreateBranch(branchName)).ConfigureAwait(false);
        }

        public async Task Handle(RequestUpdateFile message, IMessageHandlerContext context)
        {
            await this.logic.UpdateFile(message.BranchName, message.FileName, message.Content).ConfigureAwait(false);
            await context.Reply(new ResponseUpdateFile()).ConfigureAwait(false);
        }

        public async Task Handle(RequestCreatePullRequest message, IMessageHandlerContext context)
        {
            var pullRequestUri = await this.logic.CreatePullRequest(message.BranchName).ConfigureAwait(false);
            await context.Reply(new ResponseCreatePullRequest(pullRequestUri)).ConfigureAwait(false);
        }
    }    
}

File: src/endpoint/Bc.Endpoint/GitHubPullRequestCreation.fs
module Bc.GitHubPullRequestCreation

open System
open System.Configuration
open System.Threading.Tasks
open Bc.Contracts.Internals.Endpoint.GitHubPullRequestCreation.Logic
open Bc.Contracts.Internals.Endpoint.GitHubPullRequestCreation.Messages
open GitHubApi
open NServiceBus
open NServiceBus.Persistence.Sql

module ConfigurationProvider =

    let isUseFakes =
        Convert.ToBoolean(ConfigurationManager.AppSettings.["IsUseFakes"])

module GitHub =
    let createBranch (creationDate: DateTime) =
        async {
            match ConfigurationProvider.isUseFakes with
            | true ->
                return "branch_1234"
            | false ->
                let branchName = sprintf "c-%s" (creationDate.ToString("yyyy-MM-dd-HH-mm-ss-fff"))
                do! GitHubApi.CreateRepositoryBranch.execute
                        GitHubConfigurationProvider.userAgent
                        GitHubConfigurationProvider.authorizationToken
                        GitHubConfigurationProvider.repositoryName
                        GitHubConfigurationProvider.masterBranchName
                        branchName
                return branchName
        }

    let updateFile fileName branchName content =
        async {
            match ConfigurationProvider.isUseFakes with
            | true ->
                ()
            | false ->
                do! GitHubApi.UpdateFile.execute
                                            GitHubConfigurationProvider.userAgent
                                            GitHubConfigurationProvider.authorizationToken
                                            GitHubConfigurationProvider.repositoryName
                                            fileName
                                            branchName
                                            content
        }

    let createPullRequest branchName =
        async {
            match ConfigurationProvider.isUseFakes with
            | true ->
                return "uri_1234"
            | false ->
                let! pullRequestUri = GitHubApi.CreatePullRequest.execute
                                                                    GitHubConfigurationProvider.userAgent
                                                                    GitHubConfigurationProvider.authorizationToken
                                                                    GitHubConfigurationProvider.repositoryName
                                                                    branchName
                                                                    GitHubConfigurationProvider.masterBranchName
                return pullRequestUri
        }

type PolicyData () =
    inherit ContainSagaData()

    member val CommentId = Guid.Empty with get, set
    member val FileName = "" with get, set
    member val Content = "" with get, set
    member val AddedDate = DateTime.MinValue with get, set
    member val BranchName = "" with get, set


[<SqlSaga(nameof Unchecked.defaultof<PolicyData>.CommentId)>]
type Policy() =
    inherit Saga<PolicyData>()
        override this.ConfigureHowToFindSaga(mapper: SagaPropertyMapper<PolicyData>) =
            mapper.MapSaga(fun saga -> saga.CommentId :> obj)
                  .ToMessage<RequestCreateGitHubPullRequest>(fun message -> message.CommentId :> obj) |> ignore

    interface IAmStartedByMessages<RequestCreateGitHubPullRequest> with
        member this.Handle(message, context) =
            this.Data.FileName <- message.FileName
            this.Data.Content <- message.Content
            this.Data.AddedDate <- message.AddedDate

            context.Send(RequestCreateBranch(this.Data.AddedDate))

    interface IHandleMessages<ResponseCreateBranch> with
        member this.Handle(message, context) =
            this.Data.BranchName <- message.BranchName

            context.Send(RequestUpdateFile(this.Data.BranchName, this.Data.FileName, this.Data.Content))

    interface IHandleMessages<ResponseUpdateFile> with
        member this.Handle(message, context) =
            context.Send(RequestCreatePullRequest(this.Data.BranchName))

    interface IHandleMessages<ResponseCreatePullRequest> with
        member this.Handle(message, context) =
            this.MarkAsComplete()
            this.ReplyToOriginator(
                context,
                ResponseCreateGitHubPullRequest(this.Data.CommentId, message.PullRequestUri))

type PolicyHandlers() =
    interface IHandleMessages<RequestCreateBranch> with
        member this.Handle(message, context) =
            async {

                let! branchName = GitHub.createBranch message.AddedDate
                do! context.Reply(ResponseCreateBranch(branchName)) |> Async.AwaitTask

            } |> Async.StartAsTask :> Task

    interface IHandleMessages<RequestUpdateFile> with
        member this.Handle(message, context) =
            async {

                do! GitHub.updateFile message.FileName message.BranchName message.Content
                do! context.Reply(ResponseUpdateFile()) |> Async.AwaitTask

            } |> Async.StartAsTask :> Task

    interface IHandleMessages<RequestCreatePullRequest> with
        member this.Handle(message, context) =
            async {

                let! pullRequestUri = GitHub.createPullRequest message.BranchName
                do! context.Reply(ResponseCreatePullRequest(pullRequestUri)) |> Async.AwaitTask

            } |> Async.StartAsTask :> Task
